075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,500 @@
+
+## Motivation
+
+The rebalancing operation during scaling up and down is a manual two-step process:
+
+* Scaling up: 
+   1. Scaling the cluster up by adding brokers
+   2. Running a rebalance
+* Scaling down: 
+   1. Running a rebalance
+   2. Scaling the cluster down by removing brokers
+
+In order to simplify the process, it would be great for the user to deal with just scaling the cluster and then having the operator taking care of running a rebalancing operation automatically when needed.
+
+## Proposal
+
+This proposal is about providing the user the possibility to automatically run a rebalancing when they scale the cluster up (by adding brokers) or scale the cluster down (by removing brokers) by changing the `spec.replicas` within a `Kafka` or `KafkaNodePool` custom resource.
+
+### Extending the `Kafka.spec.cruiseControl` field with auto-rebalancing related configuration
+
+The `Kafka` custom resource is extended by adding a new `spec.cruiseControl.autoRebalance` field.
+By default, without such field, no auto-rebalancing runs on scaling.
+
+The `spec.cruiseControl.autoRebalance` field allows to specify a list of rebalancing modes with the corresponding configuration to be used.
+For each entry in the list, the mode is defined by the `spec.cruiseControl.autoRebalance.mode` field (`add-brokers` or `remove-brokers`) and the corresponding rebalancing configuration is defined as a reference to a "template" `KafkaRebalance` custom resource (read later for more details), by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration.
+
+In the following example, a user can decide to use the same rebalancing configuration when adding or removing brokers.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the same rebalancing configuration when adding or removing brokers
+      - mode: add-brokers
+        template:
+          name: my-add-remove-brokers-rebalancing-template
+      - mode: remove-brokers
+        template:
+          name: my-add-remove-brokers-rebalancing-template
+```
+
+Furthermore, it is possible to use the default Cruise Control rebalancing configuration by omitting the corresponding field.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...
+    autoRebalance:
+      # using the default Cruise Control rebalancing configuration when adding or removing brokers
+      - mode: add-brokers
+      - mode: remove-brokers
```

Comment:
What about applying the same logic for a config like `autoRebalance: {}`?

Reply:
I think it's a bit different. If your `KafkaRebalance.spec` is `{ }` you are implicitly specifying a `full` rebalancing with default CC config. Here you would specify all available modes with CC default. But then just a missing mode what would mean? I think that as we went for disabled by default and you need to specify modes, it's better to be explicit same way here.

Reply:
> But then just a missing mode what would mean?

Would mean the same: defaults are applied for the missing mode with an info level message. TBH, I don't see why one would need "add" but not "remove" or vice versa. Instead I see the need to have different rebalance setup for "add" and "remove".

When possible, I see a lot of value in having a simpler configuration.

Reply:
> Would mean the same: defaults are applied for the missing mode with an info level message

No because, missing mode means ... no mode activated. It's disabled by default.

Reply:
Sorry if I insist, but I would like to understand what's the use case to enable only one mode.

Reply:
I think it's just a matter of giving freedom. Also, while autorebalancing on scale down is kind of mandatory to scale, for scale up maybe user would use the new brokers just for new topics and not having the autorebalancing since the beginning. I don't want to be too strict, we don't know how users are going to use this new feature.

Reply:
Ok, this makes sense, can you please add to the proposal?

Reply:
I added it. A few lines below, after the latest example.